### ISSUE TITLE
[Other] Switch primary data structure from nested array to object

### ISSUE THREAD

**Issue by justinkambic** [elastic/synthetics-recorder#169]
### Task summary

Today, we store journey data in a custom intermediate representation that implements a type `ActionContext[][]`. This has worked great for us thus far, because we can do powerful mutations on the top-level array to get things done with little code.

Unfortunately, we have arrived at a point where we desire to attach stateful data to the nested arrays. For example, we want to let users [give custom names](https://github.com/elastic/synthetics-recorder/issues/64) to steps (i.e. the nested arrays).

I've explored the option of adding custom fields to an extension of the `Array` prototype, but this is problematic. `JSON` doesn't allow non-numeric keys on `Array`s, so any time our data structure is subjected to `stringify`, our custom fields are dropped on the floor. Likewise, some kind of custom mapping function that creates a sidecar array with the custom step fields then creates added complexity and novelty in upstream code, like the generator code in Synthetics.

I've avoided the option of re-implementing this data structure from the ground up, but for the functionality we need I think it is the best choice. I am open to other ideas, however.

**Comment by kyungeunni** [comment]
I'm in favour of amending the data structure within the recorder as it makes things much simpler.
I just imagined data structure like this:

```ts
type Steps = Step[];
interface Step {
  name: string;
  actions: ActionContext[];
  // and all the other fields we would want to add in the future.. 
}
```
In this case, when interacting with @elastic/synthetics as we just need to serialize it with `steps.map(step => step.actions)` when generating the code and it's compatible with the previous model.

**Comment by vigneshshanmugam** [comment]
Bit of context, The reason why I went with flat Array instead of a Map when designing the IR was pretty much to have interoperability between the actions emitted by the Playwright and the one we maintain which can be flattened and 
reconstructed at a much easier phase. 

>Likewise, some kind of custom mapping function that creates a sidecar array with the custom step fields then creates added complexity and novelty in upstream code, like the generator code in Synthetics.

I am not opposed to changing the structure to Map now, But keep in mind the other seamless optimizations that we do for diffing and assertion would become a lot trickier with Map than with the flattened structure. 

Could you put up a proposal or a glue code on your thoughts?

**Comment by lucasfcosta** [comment]
I really like @kyungeunni, it's very similar to what I initially had in mind when I saw this issue. IMO trying to extend `Array` tends to be problematic. [I tried to do that myself for infinite arrays in the past and it can be a bit tricky](https://github.com/lucasfcosta/harray).

In general I'd just prefer to have arrays of objects with a nested property containing yet another array (as @kyungeunni). Then extracting and flattening that should be quite easy. We could even have a few utils for dealing with objects of that "class".
